jina/peapods/runtime.py

```diff
@@ -0,0 +1,180 @@
+
+from .zmq import send_ctrl_message, Zmqlet
+from ..enums import PeaRoleType
+from ..excepts import PeaFailToStart
+
+from .. import __ready_msg__, __stop_msg__
+from ..helper import typename
+from ..logging import JinaLogger
+from . import Pea
+
+__all__ = ['RuntimeMeta', 'RunTime']
+
+
+def _get_event(obj: 'RunTimeSupport') -> Event:
+    if isinstance(obj, threading.Thread):
+        return threading.Event()
+    elif isinstance(obj, multiprocessing.Process):
+        return multiprocessing.Event()
+    else:
+        raise NotImplementedError
+
+
+def _make_or_event(obj: 'RunTimeSupport', *events) -> Event:
+    or_event = _get_event(obj)
+
+    def or_set(self):
+        self._set()
+        self.changed()
+
+    def or_clear(self):
+        self._clear()
+        self.changed()
+
+    def orify(e, changed_callback):
+        e._set = e.set
+        e._clear = e.clear
+        e.changed = changed_callback
+        e.set = lambda: or_set(e)
+        e.clear = lambda: or_clear(e)
+
+    def changed():
+        bools = [e.is_set() for e in events]
+        if any(bools):
+            or_event.set()
+        else:
+            or_event.clear()
+
+    for e in events:
+        orify(e, changed)
+    changed()
+    return or_event
+
+
+class RuntimeMeta(type):
+    """Meta class of :class:`BasePea` to enable switching between ``thread`` and ``process`` backend. """
+    _dct = {}
+
+    def __new__(cls, name, bases, dct):
+        _cls = super().__new__(cls, name, bases, dct)
+        RuntimeMeta._dct.update({name: {'cls': cls,
+                                        'name': name,
+                                        'bases': bases,
+                                        'dct': dct}})
+        return _cls
+
+    def __call__(cls, *args, **kwargs) -> 'RuntimeMeta':
+        # switch to the new backend
+        _cls = {
+            'thread': threading.Thread,
+            'process': multiprocessing.Process,
+        }.get(getattr(args[0], 'runtime', 'thread'))
+
+        # rebuild the class according to mro
+        for c in cls.mro()[-2::-1]:
+            arg_cls = RuntimeMeta._dct[c.__name__]['cls']
+            arg_name = RuntimeMeta._dct[c.__name__]['name']
+            arg_dct = RuntimeMeta._dct[c.__name__]['dct']
+            _cls = super().__new__(arg_cls, arg_name, (_cls,), arg_dct)
+
+        return type.__call__(_cls, *args, **kwargs)
+
+
+class RunTime(metaclass=RuntimeMeta):
+    def __init__(self, args: Union['argparse.Namespace', Dict]):
+        self.args = args
+        self.name = self.__class__.__name__
+
+        self.is_ready_event = _get_event(self)
+        self.is_shutdown = _get_event(self)
+        self.ready_or_shutdown = _make_or_event(self, self.is_ready_event, self.is_shutdown)
+        self.is_shutdown.clear()
+
+        if isinstance(self.args, argparse.Namespace):
+            if self.args.name:
+                self.name = f'support-{self.args.name}'
+            elif self.args.role == PeaRoleType.PARALLEL:
```

Comment:
Why do we have anything related to `Peas` in the class `Runtime`? Things like `Remote`, `Client` don't need to be Peas.

Reply:
yes, very early stage